Dockerfile.rocm

```diff
@@ -39,79 +40,155 @@ RUN apt-get update && apt-get install -y \
     build-essential \
     wget \
     unzip \
-    nvidia-cuda-toolkit \
     tmux \
     ccache \
  && rm -rf /var/lib/apt/lists/*
 
-### Mount Point ###
-# When launching the container, mount the code directory to /app
+# When launching the container, mount the code directory to /vllm-workspace
 ARG APP_MOUNT=/vllm-workspace
-VOLUME [ ${APP_MOUNT} ]
 WORKDIR ${APP_MOUNT}
 
-RUN python3 -m pip install --upgrade pip
-RUN python3 -m pip install --no-cache-dir fastapi ninja tokenizers pandas
+RUN pip install --upgrade pip
+# Install torch == 2.4.0 on ROCm
+RUN case "$(ls /opt | grep -Po 'rocm-[0-9]\.[0-9]')" in \
+        *"rocm-5.7"*) \
+            pip uninstall -y torch \
+            && pip install --no-cache-dir --pre torch==2.4.0.dev20240612 \
+               --index-url https://download.pytorch.org/whl/nightly/rocm5.7;; \
```

Comment:
Do you know whether base image rocm/pytorch:rocm6.1.2_ubuntu20.04_py3.9_pytorch_staging works for pytorch rocm5.7 and rocm 6.0 pytorch wheels?

Reply:
Not sure if I understand this question, or the spirit behind the question: is the question whether "pytorch_staging" works on ROCm 5.7 or 6.0 without wheels, or whether these wheels work on a non-"pytorch_staging" image? Because you'd never install rocm5.7 or rocm6.0 wheels on a rocm6.1.2 image.

For the first question, the answer is "no": you need at least PyTorch 2.4.0 for `torch.cuda._device_count_amdsmi`, while `pytorch_staging` goes up to PyTorch 2.3.0. For the second, the answer is "yes": the relevant wheels for each ROCm version are designed to (and work on) any official base image from the relevant versions.

Reply:
sorry for the confusing question. I was thinking about whether we should use the "staging" as the base image as the name means its testing status. And also whether rocm-6.0 + torch 2.4  or rocm-5.7 + torch 2.4, works out of box, for certain users on certain devices (especially, navi3x users, for example), because in the past the supported based docker images has torch 2.0.1 or torch 2.1.1.

Reply:
"Staging" in this case refers to the PyTorch version (2.3.0, which we are targeting for stable release sometime), so it's a moot point when we're upgrading to a PyTorch 2.4.0 nightly. And this particular nightly has not raised problems on a wide range of internal testing on multiple GPUs architectures, including MI300, MI250, and Navi32.

Let's keep in mind the broader picture here: whether this PR is an improvement to the status quo, as opposed to whether this PR is "perfect". The status quo has multiple failing tests, even on the tests AMD supports, and most importantly, does not have ROCm 6.1 support, which brings in a host of performance improvements. And upstream vLLM has moved such that PyTorch 2.3+/2.4+ on ROCm is needed for a full complement of tests.

We can offer this support: even though on paper it is "experimental", it has not raised issues in testing. Unless someone provides a specific failing example that arises as a result of this PR, quibbling about potential issues that may arise because of the "experimental" or "nightly" status, without doing the legwork to verify that such issues actually exist, seems to me to be an example of making the perfect the enemy of the good.